common/src/main/java/io/netty/util/internal/Hidden.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2019 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.util.internal;
+
+import io.netty.util.concurrent.FastThreadLocalThread;
+import reactor.blockhound.BlockHound;
+import reactor.blockhound.integration.BlockHoundIntegration;
+
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ * Contains classes that must be have public visibility but are not public API.
+ */
+class Hidden {
+
+    /**
+     * This class integrates Netty with BlockHound.
+     *
+     * It is public but only because of the ServiceLoader's limitations
+     * and SHOULD NOT be considered a public API.
+     */
+    @UnstableApi
+    @SuppressJava6Requirement(reason = "BlockHound is Java 8+, but this class is only loaded by it's SPI")
+    public static final class NettyBlockHoundIntegration implements BlockHoundIntegration {
+
+        @Override
+        public void applyTo(BlockHound.Builder builder) {
+            builder.nonBlockingThreadPredicate(new Function<Predicate<Thread>, Predicate<Thread>>() {
```

Comment:
```java
builder.allowBlockingCallsInside(
    "io.netty.channel.nio.NioEventLoop",
    "handleLoopException"
);
```

Reply:
```java
builder.allowBlockingCallsInside(
    "io.netty.util.HashedWheelTimer",
    "waitForNextTick"
);
```

Reply:
unsure about ThreadDeathWatcher as it's marked deprecated and not used in the Netty codebase

Reply:
```java
builder.allowBlockingCallsInside(
    "io.netty.util.concurrent.SingleThreadEventExecutor",
    "confirmShutdown"
);
```

Reply:
@bsideup I wonder if we could just whitelist all of `io.netty.*` or if there is support to annotate methods with something so these are ignored. Having to specify a string for all of them seems very error-prone

Reply:
@normanmaurer and would probably break with shaded Netty.

Reply:
@johnou 
Shading will work fine because it searches for the string occurrences.
I see that you whitelisted a few util methods. It may not be a good solution, we usually recommend whitelisting the "business logic" methods that call such blocking methods and cannot avoid them. Whitelisting the util method may lead to a situation where it will get called from a method that must not call any blocking methods.

@normanmaurer 
See my previous answer. If you whitelist `io.netty.*`, a user-called Netty code (some util or maybe even internal) will block user's thread but he won't notice.

Reply:
Also... don't you want to test Netty with BlockHound to double check that **you** are not blocking where you should not? 😊

Reply:
@bsideup but I think we need to whitelist as otherwise it would not work anymore... these methods are used internally by netty and can not be called by users.

Reply:
@normanmaurer 
have you considered whitelisting `run` methods instead, from where they called?
And potentially blacklisting the part where you call back into the user's code?

Reply:
@bsideup I am not sure what you suggest. Can you show me an example for one of the methods @johnou listed above ?

Reply:
@normanmaurer sure!

For example, not `NioEventLoop#handleLoopException`, but `NioEventLoop#run` and blacklist `NioEventLoop#runAllTasks` for example.
So that `run` is expected to be blocking (because it is an infinite loop) but the user code it calls must not block this loop (well, unless it can)

Reply:
I see... can this also be done via an annotation ?

Reply:
Only if you add your own annotation and scan for it later.

To give you a bit of context:
We use String-based API for the whitelist/blacklist for a few reasons:
1. to not trigger the classloading (so that we don't have to retransform the classes later on)
2. some classes are internal/private and cannot be referenced even from a project-specific integration
3. sometimes you need to mark some JDK methods that you cannot modify (e.g. add an annotation) otherwise

That said, we may consider adding some sort of an annotation scanning, although I believe it will increase the installation time of BlockHound, since we will need to scan basically everything to find such methods

Reply:
@bsideup I see... I just find using "strings" quite error-prone in practice :/

That said can you just adjust the PR to do the string stuff for now for the methods within netty that use `Thread.sleep(...)` ?

Reply:
@normanmaurer

FYI since we're talking about Netty's integration and the classes are already loaded, nothing stops you from using:
```java
builder.allowBlockingCallsInside(
    io.netty.util.concurrent.SingleThreadEventExecutor.class.getName(),
    "confirmShutdown"
);
```
(unless the class is private/internal, in that case, such package may have it's own integration)

I will update the PR and add the whitelisted methods 👍

Anything else you think is missing so far, or this is the only thing? :)

Reply:
I think that is the only thing... Once this is merged I guess we also want to run in on the CI with BlockHound enabled once a day.

Reply:
Ok, cool, will update the PR in a moment :)

Re the CI - I would even recommend having it on by default for every CI run given the very low overhead, so that the contributions are also protected from the blocking calls. 

That's of course up to you, but I would be curious to know the arguments **against** having it enabled by default. At least we plan to make it **the** tool for async Java frameworks to verify their asynchronosity ☺️

Reply:
I added all the methods except `OioSocketChannel #checkInputShutdown` (my IDE shows it as "unused" and the class is deprecated anyways)